File: HX1.py
import numpy as np
from scipy.integrate import solve_ivp
from scipy.sparse import diags, csc_matrix
from scipy.sparse.linalg import spsolve

from parameters import *
from ode_solver import ode_solver

def HX1(y_hx1, Ts_HX1_L, Tss_HX1_0, params, step):
    
    V_he_s = params['V_he_s']
    U_hx = params['U_hx']
    M_he_s = params['M_he_s']
    c_p_s = params['c_p_s']
    V_he_ss = params['V_he_ss']
    U_hx = params['U_hx']
    M_he_ss = params['M_he_ss']
    c_p_ss = params['c_p_ss']
    Nx = params['Nx']
    dx = params['dx']
    u_L = params['u_L']
    u_H = params['u_H']
    v_L = params['v_L']
    v_H = params['v_H']
    err = params['err']
    u_init = params['u_init']
    v_init = params['v_init']
    
    # New Form Parameters
    C1 = V_he_s
    C2 = -U_hx / (M_he_s * c_p_s)
    C3 = V_he_ss
    C4 = U_hx / (M_he_ss * c_p_ss)

    # Discretize the spatial domain
    # A_HX=np.diag(-np.ones(Nx))+ np.diag(np.ones(Nx-1), 1) / dx
    # # A_HX = (-2*np.diag(np.ones(Nx)) + np.diag(np.ones(Nx - 1), 1) + np.diag(np.ones(Nx - 1), -1)) / dx
    # A_HX[0, 0] = 1 / dx
    # A_HX[-1, -1] = 1 / dx
    A_HX = np.diag(-2 * np.ones(Nx)) + np.diag(np.ones(Nx-1), 1) + np.diag(np.ones(Nx-1), -1)
    A_HX[0, 0] = A_HX[-1, -1] = -1
    A_HX[0, 1] = A_HX[-1, -2] = 0
    # A_HX = A_HX / dx**2
    A_HX_sparse = csc_matrix(A_HX) / dx**2
    
    y_hx1[:Nx,-1]=Ts_HX1_L
    y_hx1[Nx:,0]=Tss_HX1_0
    
    # y is a vector with u and v concatenated
    def pde_to_ode_hx1(t,y):
        u=y[:Nx]
        v=y[Nx:]
        
        du_dt = C1 * (A_HX_sparse @ u) + C2 * (u - v) + err
        dv_dt = C3 * (A_HX_sparse @ v) + C4 * (u - v) + err
        
        # Apply time-varying boundary conditions
        du_dt[0] = u_L - u[0] / dx
        du_dt[-1] = u_H - u[-1] / dx
        dv_dt[0] = v_L - v[0] / dx
        dv_dt[-1] = v_H - v[-1] / dx
        
        dydt = np.concatenate([du_dt, dv_dt])
        return dydt
    
    # Initial condition vector
    if step==0:
        y0 = np.concatenate([u_init, v_init])
    else:
        y0 = y_hx1[:,-1]
        
    # solution_y_hx1 = solve_ivp(pde_to_ode_hx1, (step, step+1), y0, t_eval=t, method='RK45')     
    bc=[]
    solution_y_hx1 = ode_solver(y0, bc, pde_to_ode_hx1, params)
        
    # y_hx1 is a vector at time step+1
    y_hx1 = solution_y_hx1.y   
    # print("testHX1")
    # print(y_hx1.shape)
    return y_hx1

File: HX1.cpp
#include "HX1.hpp"
#include "parameters.hpp"
#include "ode_solver_HX1.hpp"

void pde_to_ode_hx1(float t, const float y[length_hx], float dydt[length_hx], Param_HX1 &params) {
    // Split the input state vector y into u and v
    const float* u = &y[0];    // u is the first half of y
    const float* v = &y[Nx];   // v is the second half of y

    float du_dt[Nx] = {0};    // Array to store the derivative of u
    float dv_dt[Nx] = {0};    // Array to store the derivative of v

    // Compute du_dt and dv_dt using the provided constants
    for (int i = 0; i < Nx; ++i) {
#pragma HLS UNROLL
        for (int idx = params.A_HX_csr.row_pointers[i]; idx < params.A_HX_csr.row_pointers[i + 1]; ++idx) {
#pragma HLS UNROLL
            int j = params.A_HX_csr.col_indices[idx];
            du_dt[i] += params.A_HX_csr.values[idx] * u[j];
            dv_dt[i] += params.A_HX_csr.values[idx] * v[j];
        }
        du_dt[i] += params.C2_1 * (u[i] - v[i]);
        dv_dt[i] += params.C4_1 * (u[i] - v[i]);
        dydt[i] = du_dt[i];        // First half for du_dt
        dydt[Nx + i] = dv_dt[i];   // Second half for dv_dt
    }

    // Apply time-varying boundary conditions
    dydt[0] = params.u_L - dydt[0];
    dydt[Nx - 1] = params.u_H - dydt[Nx - 1];
    dydt[Nx] = params.v_L - dydt[Nx];
    dydt[2*Nx - 1] = params.v_H - dydt[2*Nx - 1];
}

void HX1(float y_hx1[length_hx], float Ts_HX1_L, int step) {
    Param_HX1 params;

    // Initial condition vector y0
    float y0[length_hx];
    if (step == 0) {
        for (int i = 0; i < Nx; ++i) {
#pragma HLS UNROLL
            y0[i] = params.u_init[i];
            y0[Nx + i] = params.v_init[i];
        }
    } else {
        for (int i = 0; i < length_hx; ++i) {
#pragma HLS UNROLL
            y0[i] = y_hx1[i];
        }
    }
    y0[Nx - 1] = Ts_HX1_L;

    // Solve the system of ODEs
    ode_solver_hx1(y0, pde_to_ode_hx1, step, params);

    // Update y_hx1 with the solution from the ODE solver
    for (int i = 0; i < length_hx; ++i) {
#pragma HLS UNROLL
        y_hx1[i] = y0[i];  // Update the original array with the new values
    }
}

